### PR TITLE
fix: #23230 solutions from centrifuge are now layered and are not poured from vial all at the same time

### PR DESCRIPTION
**Pull request by Fildrance**
## About the PR
Centrifuge will now force soutions in vial to pour reagents by pulling them from container as from stack, exausting them one by one, until requested amount is pulled (instead of mixing them all togather). Extracting all reagents from vial, or adding any amount of other solution into it will break this effect.

## Why / Balance
Fixing #23230, making centrifuge a bit more viable, while maintaining focus on ghetto chem.

## Technical details
**Solution** will now have _IsSeparatedByLayers_ flag. Centrifuge run will set it 'true', adding liquids and empty-ing will drop it. Flag is not copied on clone. On _SplitSolution_ flag forces **Solution** to get not 'part' of each reagent, but to drain them one by one from stack until requested amount will be acquired. **SolutionContainerMixerComponent** now have _canSeparate_ flag to enable/disable this behaviour (as centrifuge and electrolysis machine share system).

## Media
[centrifuge-separate-solution-layers.webm](https://github.com/space-wizards/space-station-14/assets/14752842/f38901a3-53f6-48f4-9450-5ec4a419a89b)

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

:cl:
- tweak: Centrifuge now separates any liquids it processes into layers

### CODE REVIEW

**Comment by Fildrance** [comment]
Non-obvious problem with this solution is - **chemical analysis goggles output reagents in sorted order, but container of reagents in solution is just stack**. So order of poured liquids is not same as order displayed in hint, and it cannot be.
Overall it still makes centrifuge more viable, but with binding on vials and their 30u max amount its still seems dead. Problem with goggles echances it. I can change them to display stack that is inside, and not ordered list, but it can bring more harm then good (potentially), as it'll be much more difficult to compare ingredient lists.

I have branch where i created simplified chem-master-like UI for centrifuge (please look at https://github.com/space-wizards/space-station-14/issues/23230#issuecomment-1917946279) but it still won't be enough by itself to make centrifuge important part of gameplay, so i'm more then happy to merge current PR and throw that UI branch into abyss.

File: Content.Shared/Chemistry/Components/Solution.cs [hunk 1]
```diff
@@ -27,6 +27,9 @@ public sealed partial class Solution : IEnumerable<ReagentQuantity>, ISerializat
         [ViewVariables]
         public FixedPoint2 Volume { get; set; }
 
+        [ViewVariables(VVAccess.ReadWrite)]
+        public bool IsSeparatedByLayers { get; set; }
```
Comment (0x6273) [inline Content.Shared/Chemistry/Components/Solution.cs:31]: This has to be a datafield

File: Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerMixerSystem.cs [hunk 2]
```diff
@@ -102,6 +102,8 @@ public void FinishMix(Entity<SolutionContainerMixerComponent> entity)
                 continue;
 
             _solution.UpdateChemicals(soln.Value, true, reactionMixer);
+            if(comp.CanSeparate)
```
Comment (0x6273) [inline Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerMixerSystem.cs:105]:
```suggestion
            if (comp.CanSeparate)
```

**Comment by moonheart08** [comment]
Instead of adding a new flag to the existing component, make a new component and introduce events as needed to implement the new behavior in a second system.

**Review by 0x6273 — CHANGES_REQUESTED** [review]

File: Content.Shared/Chemistry/Components/Solution.cs (hunk 1 above)
Comment (Fildrance) [reply Content.Shared/Chemistry/Components/Solution.cs:32]: fixed

File: Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerMixerSystem.cs (hunk 2 above)
Comment (Fildrance) [reply Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerMixerSystem.cs]: fixed

**Comment by Fildrance** [comment]
Class names and idea of introducing system to flip one bool flag seems ify, but otherwise seems ok. Any other advices? Maybe on new system/event naming?

**Comment by Beridot** [comment]
I love this!  I can't WAIT for it to be merged.. :)

**Comment by Beridot** [comment]
> I love this! I can't WAIT for it to be merged.. :)
> Forgotten to time..

**Comment by 0x6273** [comment]
There should be visual indication that the solution in the vial is layered, by drawing each reagent separately instead of mixing the colors. Right now a vial that is layered looks identical to one that isn't.

**Comment by 0x6273** [comment]
Also, the order that reagents are poured out should be based on their density. Chem goggles should list the reagents in order of density, but only for layered solutions.

**Comment by Tayrtahn** [comment]
> There should be visual indication that the solution in the vial is layered, by drawing each reagent separately instead of mixing the colors. Right now a vial that is layered looks identical to one that isn't.

I agree that it would be great to see the laying, but that doesn't really seem viable with the way that fill level visuals are implemented.

File: Content.Shared/Chemistry/Components/Solution.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Shared/Chemistry/Components/Solution.cs:32]: This needs documenting so it's clear what it even does.

File: Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerMixerSystem.cs [hunk 3]
```diff
@@ -102,6 +102,8 @@ public void FinishMix(Entity<SolutionContainerMixerComponent> entity)
                 continue;
 
             _solution.UpdateChemicals(soln.Value, true, reactionMixer);
+            var mixerMixedEvent = new SolutionContainerMixerMixedEvent(soln.Value);
+            RaiseLocalEvent(container.Owner, ref mixerMixedEvent, true);
```
Comment (metalgearsloth) [inline Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerMixerSystem.cs:106]: Why raise it as broadcast?

File: Content.Shared/Chemistry/Components/SolutionSeparatorComponent.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+using Content.Shared.Chemistry.EntitySystems;
+
+namespace Content.Shared.Chemistry.Components;
+
+/// <summary>
+/// Marker that machine can separate solution that it works with into layers,
+/// so it will pour out reagent by reagent and not all mixed up.
+/// </summary>
+/// <seealso cref="SharedSolutionContainerMixerSystem"/>
+[RegisterComponent]
```
Comment (metalgearsloth) [inline Content.Shared/Chemistry/Components/SolutionSeparatorComponent.cs:11]:
```suggestion
/// <seealso cref="SharedSolutionContainerMixerSystem"/>
[RegisterComponent, NetworkedComponent]
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Think I agree on making it a new component to avoid bloating solutions more.

File: Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerMixerSystem.cs (hunk 3 above)
Comment (Fildrance) [reply Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerMixerSystem.cs]: fixed

**Comment by AJCM-git** [comment]
@Fildrance are you coming back to this?

**Comment by Fildrance** [comment]
> @Fildrance are you coming back to this?

To be honest i don't really want it - as it is now. Maybe after someone fixes layering and displaying layers in vials/container? I don't think this will happen either, but if it will - i'll fix other stuff :< for now i will be busy with discounts, i think. We can close it for a time.

**Pull request closed without merging**